pkg/operator/controller/ingress/serviceca_configmap.go

```diff
@@ -0,0 +1,104 @@
+)
+
+// ensureServiceCAConfigMap ensures the configmap for the service CA bundle
+// exists.  Returns a Boolean indicating whether the configmap exists, the
+// configmap if it does exist, and an error value.
+func (r *reconciler) ensureServiceCAConfigMap() (bool, *corev1.ConfigMap, error) {
+	wantCM, desired, err := desiredServiceCAConfigMap()
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to build configmap: %v", err)
+	}
+
+	haveCM, current, err := r.currentServiceCAConfigMap()
+	if err != nil {
+		return false, nil, err
+	}
+
+	switch {
+	case !wantCM && !haveCM:
+		return false, nil, nil
+	case !wantCM && haveCM:
+		if err := r.client.Delete(context.TODO(), current); err != nil {
+			if !errors.IsNotFound(err) {
+				return true, current, fmt.Errorf("failed to delete configmap: %v", err)
+			}
+		} else {
+			log.Info("deleted configmap", "configmap", current)
+		}
+	case wantCM && !haveCM:
+		if err := r.client.Create(context.TODO(), desired); err != nil {
+			return false, nil, fmt.Errorf("failed to create configmap: %v", err)
+		}
+		log.Info("created configmap", "configmap", desired)
+	case wantCM && haveCM:
+		if updated, err := r.updateServiceCAConfigMap(current, desired); err != nil {
+			return true, nil, fmt.Errorf("failed to update configmap: %v", err)
+		} else if updated {
+			log.Info("updated configmap", "configmap", desired)
+		}
+	}
+
+	return r.currentServiceCAConfigMap()
+}
+
+// desiredServiceCAConfigMap returns the desired configmap for the service CA
+// bundle.  Returns a Boolean indicating whether a configmap is desired, as well
+// as the configmap if one is desired.
+func desiredServiceCAConfigMap() (bool, *corev1.ConfigMap, error) {
+	name := controller.ServiceCAConfigMapName()
+	cm := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"description": "ConfigMap providing service CA bundle.",
+				"service.beta.openshift.io/inject-cabundle": "true",
+			},
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+	}
+
+	return true, &cm, nil
+}
+
+// currentServiceCAConfigMap returns the current configmap for the service CA
+// bundle.  Returns a Boolean indicating whether the configmap existed, the
+// configmap if it did exist, and an error value.
+func (r *reconciler) currentServiceCAConfigMap() (bool, *corev1.ConfigMap, error) {
+	cm := &corev1.ConfigMap{}
+	if err := r.client.Get(context.TODO(), controller.ServiceCAConfigMapName(), cm); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, cm, nil
+}
+
+// updateServiceCAConfigMap updates the configmap for the service CA bundle.
+// Returns a Boolean indicating whether the configmap was updated, and an error
+// value.
+func (r *reconciler) updateServiceCAConfigMap(current, desired *corev1.ConfigMap) (bool, error) {
+	if current.Annotations["service.beta.openshift.io/inject-cabundle"] == "true" {
```

Comment:
This acts as a gate? Once present we never update - is that the implied semantics?  If so, maybe the check should occur by callers of `updateServiceCAConfigMap` so that update always means update.

Reply:
That is pretty much right: cluster-ingress-operator manages the configmap's lifecycle and annotation, and the serving cert signer manages the data.  If the configmap's annotation were deleted, cluster-ingress-operator would restore it, but otherwise it should not update the configmap.  Would it be best to fold `updateServiceCAConfigMap` into `ensureServiceCAConfigMap`?

Reply:
For me this was mostly about the semantics. A function called `update...` doesn't always update, but only if you peek at its implementation. If it is only called from one place them maybe folding would help.

Reply:
Here, "updated" means "update if needed", and in this particular case, "if needed" means "if the critical annotation is not set properly".  It is kind of a degenerate case of a pattern we use elsewhere in the operator, where `ensureFoo` calls `createFoo`, `deleteFoo`, and `updateFoo`, and `updateFoo` determines whether update is needed.

Reply:
OK, fine for me. Thanks for explaining the pattern.